scripts/tools/generate_nrfconnect_chip_factory_data.py

```diff
@@ -0,0 +1,358 @@
+
+
+OUT_DIR = os.path.dirname(os.path.realpath(__file__))
+TOOLS = {}
+
+
+def make_dict(input_list: list):
+    names = list()
+    attrs = list()
+    for entry in input_list:
+        names.append(entry[0])
+        attrs.append(entry[1])
+    return dict(zip(names, attrs))
+
+
+def check_tools_exists():
+    TOOLS['spake2p'] = shutil.which('spake2p')
+    if TOOLS['spake2p'] is None:
+        log.error('spake2p not found, please build and add spake2p path to PATH environment variable')
+        log.error('You can find spake2p in connectedhomeip/src/tools/spake2p directory')
+        sys.exit(1)
+
+
+def gen_spake2p_params(passcode):
+    iter_count_max = 10000
+    salt_len_max = 32
+
+    cmd = [
+        TOOLS['spake2p'], 'gen-verifier',
+        '--iteration-count', str(iter_count_max),
+        '--salt-len', str(salt_len_max),
+        '--pin-code', str(passcode),
+        '--out', '-',
+    ]
+
+    output = subprocess.check_output(cmd)
+    output = output.decode('utf-8').splitlines()
+    return dict(zip(output[0].split(','), output[1].split(',')))
+
+
+def bytes_needed(n):
+    if n == 0:
+        return 1
+    return int(logarithm(n, 256)) + 1
+
+
+def convert_to_bytes(value: any):
+    if type(value) == int:
+        return value.to_bytes(bytes_needed(value), byteorder='big', signed=False)
+    elif type(value) == str:
+        return value.encode("utf-8")
+    elif type(value) == bytes:
+        return value
+    else:
+        return value
+
+
+class ValidatorError(Exception):
+    def __init__(self, message="str"):
+        self.message = message
+
+    def __str__(self):
+        return self.message
+
+
+class PartitionCreator:
+    def __init__(self, offset: int, length: int) -> None:
+        self.__ih = IntelHex()
+        self.__length = length
+        self.__offset = offset
+        self.__data_ready = False
+
+    @staticmethod
+    def encrypt(data: bytes):
+        return data
+
+    @staticmethod
+    def decrypt(data: bytes):
+        return data
+
+    def create_hex(self, data: bytes):
+        if len(data) > self.__length:
+            log.error("generated CBOR file exceeds declared maximum partition size! {} > {}".format(len(data), self.__length))
+        self.__ih.putsz(self.__offset, self.encrypt(data))
+        self.__ih.write_hex_file(OUT_DIR + "/output.hex", True)
+        self.__data_ready = True
+        return True
+
+    def create_bin(self):
+        if not self.__data_ready:
+            log.error("Please create hex file first!")
+            return False
+        self.__ih.tobinfile(OUT_DIR + "/output.bin")
+        return True
+
+
+class FactoryDataGenerator:
+    def __init__(self, arguments) -> None:
+        self.__args = arguments
+        self.__factory_data = list()
+        self.__json_data = None
+
+        try:
+            self.__validate_args()
+        except ValidatorError as e:
+            log.error(e)
+            sys.exit(-1)
+
+    def __validate_args(self):
```

Comment:
You're right. That is an excellent point!

Reply:
Hmm, I think we should check dependencies between input arguments. E.g. spake2_gen needs a passcode, and it is worth informing the user about that. Moreover, your idea is brilliant to write schema and parse it after creating a JSON file. It may improve documenting performance too.